File: Metaflow/FeatureClient.fs
namespace Metaflow

open System
open Metaflow
open Orleans
open FSharp.Control.Tasks
open Serilog

type FeatureClient(clusterClient: IClusterClient, workflows: Workflow seq) =
    let workflowMap =
        workflows
        |> Seq.map (fun f -> ((f.Feature.Model.FullName, f.Feature.Operation), f))
        |> Map.ofSeq

    member this.Delete<'model>(aggregateRootId: string, awaitState: bool) =
        task {
            let m = workflowMap
            let workflowOption =
                m
                |> Map.tryFind (typeof<'model>.FullName, Operation.DELETE)

            return!
                match workflowOption with
                | None ->
                    task {
                        return
                            { FeatureResult = NotFound
                              StepError = None }
                    }
                | Some workflow ->
                    let call =
                        { Feature = workflow.Feature
                          Input = Id(OwnedValueId { AggregateRootId = aggregateRootId })
                          AwaitState = awaitState }

                    task {
                        return!
                            Workflows.run<unit, Delete, 'model>
                                workflow
                                call
                                clusterClient
                                { RequestId = Guid.NewGuid().ToString() }
                                Log.Logger
                    }
        }

File: Metaflow.Orleans.Grains/FeatureClient.cs
using System;
using System.Threading.Tasks;
using Microsoft.FSharp.Core;
using Orleans;

namespace Metaflow.Orleans
{
    public class FeatureClient
    {
        private readonly IClusterClient _clusterClient;

        public FeatureClient(IClusterClient clusterClient)
        {
            _clusterClient = clusterClient;
        }

        public Task<FSharpResult<Unit, WorkflowFailure>> Delete<TModel>(string aggregateRootId, bool awaitState)
        {
            var grain =
                _clusterClient.GetGrain<IFeatureClientGrain>(Guid.NewGuid().ToString());

            return grain.Delete<TModel>(aggregateRootId, awaitState);
        }
    }
}